# Explore here

In [2]:
# Your code here

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error,r2_score
from scipy.stats import randint
from pickle import dump


In [4]:
Direccion_guardado = "/workspaces/Marcos-J-Gomes_RandomForest/data/processed"

total_data= pd.read_csv("https://breathecode.herokuapp.com/asset/internal-link?id=930&path=diabetes.csv")
X_train_sel_CON_outliers= pd.read_csv(f"{Direccion_guardado}/X_train_sel_CON_outliers_clean.csv")
X_test_sel_CON_outliers= pd.read_csv(f"{Direccion_guardado}/X_test_sel_CON_outliers_clean.csv")
X_train_sel_SIN_outliers=pd.read_csv(f"{Direccion_guardado}/X_train_sel_SIN_outliers_clean.csv")
X_test_sel_SIN_outliers=pd.read_csv(f"{Direccion_guardado}/X_test_sel_SIN_outliers_clean.csv")
y_train = pd.read_excel(f"{Direccion_guardado}/y_train.xlsx")
y_test  = pd.read_excel(f"{Direccion_guardado}/y_test.xlsx")

target= "Outcome"

In [17]:
datasets = {
    "CON_outliers": (X_train_sel_CON_outliers.drop(target, axis = 1), X_test_sel_CON_outliers.drop(target, axis = 1)),
    "SIN_outliers": (X_train_sel_SIN_outliers.drop(target, axis = 1), X_test_sel_SIN_outliers.drop(target, axis = 1))
}

Resultados = {}

for dataset_, (X_train, X_test) in datasets.items():
    model = RandomForestClassifier(n_estimators = 50, random_state= 17)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados[dataset_] = {
        "Model" : model,
        "Accuracy" : accuracy_score(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred)),
        
    }

mejor_dataset = max(Resultados, key=lambda x: Resultados[x]["Accuracy"])
mejor_Acurracy = Resultados[mejor_dataset]["Accuracy"]
mejor_modelo = Resultados[mejor_dataset]["Model"]

print("Resultados por dataset:")
for k, v in Resultados.items():
    print(f"{k:25s} -> Acurracy: {v["Accuracy"]:.5f} | RMSE: {v["RMSE"]:.5f}")

Resultados por dataset:
CON_outliers              -> Acurracy: 0.75325 | RMSE: 0.49674
SIN_outliers              -> Acurracy: 0.74026 | RMSE: 0.50965


Notamos una leve disminución con respecto al modelo base del "Arbol de decision" del proyecto anterior, cuyos datos eran  
CON_outliers-> Acurracy: 0.75974  
SIN_outliers-> Acurracy: 0.75325  

Esto se puede deber a que usamos un valor de "n_stimators" por debajo de defaul que es 100, haremos pruebas para intentar mejorar el Acurracy

In [22]:
dump(mejor_modelo, open("../models/random_forest_regressor_n_estimators-50_RS-17.sav", "wb"))

Pruebas

In [28]:
Resultados_2 = {}

for dataset_, (X_train, X_test) in datasets.items():
    model = RandomForestClassifier(n_estimators = 100, random_state= 17)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados_2[dataset_] = {
        "Model" : model,
        "Accuracy" : accuracy_score(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred)),
        
    }

mejor_dataset_2 = max(Resultados_2, key=lambda x: Resultados_2[x]["Accuracy"])
mejor_Acurracy_2 = Resultados_2[mejor_dataset]["Accuracy"]
mejor_modelo_2 = Resultados_2[mejor_dataset]["Model"]

print("Resultados por dataset:")
for k, v in Resultados_2.items():
    print(f"{k:25s} -> Acurracy: {v["Accuracy"]:.5f} | RMSE: {v["RMSE"]:.5f}")

Resultados por dataset:
CON_outliers              -> Acurracy: 0.75325 | RMSE: 0.49674
SIN_outliers              -> Acurracy: 0.75325 | RMSE: 0.49674


En este caso ya notamos una mejora con respecto al modelo base del Decision Tree

In [25]:
dump(mejor_modelo_2, open("../models/random_forest_regressor_n_estimators-100_RS-27.sav", "wb"))

In [29]:
Resultados_3 = {}

for dataset_, (X_train, X_test) in datasets.items():
    model = RandomForestClassifier(n_estimators = 200, random_state= 17)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados_3[dataset_] = {
        "Model" : model,
        "Accuracy" : accuracy_score(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred)),
        
    }

mejor_dataset_3 = max(Resultados_3, key=lambda x: Resultados_3[x]["Accuracy"])
mejor_Acurracy_3 = Resultados_3[mejor_dataset]["Accuracy"]
mejor_modelo_3 = Resultados_3[mejor_dataset]["Model"]

print("Resultados por dataset:")
for k, v in Resultados_3.items():
    print(f"{k:25s} -> Acurracy: {v["Accuracy"]:.5f} | RMSE: {v["RMSE"]:.5f}")

Resultados por dataset:
CON_outliers              -> Acurracy: 0.75325 | RMSE: 0.49674
SIN_outliers              -> Acurracy: 0.75974 | RMSE: 0.49016


In [26]:
dump(mejor_modelo_3, open("../models/random_forest_regressor_n_estimators-200_RS-37.sav", "wb"))

Notamos que al subir el valor de de "n_estimators" nuestro Acurracy aumenta en uno u otro de los modelos, por lo que para alcanzar un modelo con mayor precision deberiamos aumentarlo en la medida de lo posible